### PR TITLE
[IMP] hr expense: attach document from mobile/web directly using attach button as an action button on status bar

### PR DESCRIPTION
**Pull request by jpr-odoo**
Description of the issue/feature this PR addresses:
Task ID: https://www.odoo.com/web?#id=36712&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720
Pad : 

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_expense/static/src/js/form_controller.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+     * @param {OdooEvent} ev
+     */
+    _onAttachClicked: function (ev) {
+        ev.stopPropagation();
+        var self = this,
+            attrs = ev.data.attrs;
+        if (attrs.special === 'attachdocument') {
+            var canBeSaved = self.canBeSaved(ev.data.record.id);
+            if (canBeSaved && !ev.data.record.res_id) {
+                self.do_warn(_t('Warning : need to save your record for attach document!'));
```
Comment (bst-odoo) [inline addons/hr_expense/static/src/js/form_controller.js]: need to rewrite warning message in proper english

File: addons/hr_expense/static/src/js/attachment.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+odoo.define('hr_expense.Attachment', function (require) {
+"use static";
+
+var Mixins = require('web.mixins');
+var Widget = require('web.Widget');
+
+var AttachDocument = Widget.extend(Mixins.EventDispatcherMixin, {
```
Comment (ged-odoo) [inline addons/hr_expense/static/src/js/attachment.js]: Mixins.EventDispatcherMixin is already in Widget. No need to add it here

File: addons/hr_expense/views/hr_expense_views.xml [hunk 3]
```diff
@@ -28,7 +28,10 @@
             <field name="arch" type="xml">
                 <form string="Expenses" class="o_expense_form">
                 <header>
-                    <button name="submit_expenses" states="draft" string="Submit to Manager" type="object" class="oe_highlight o_expense_submit"/>
+                    <button string="Attach Document" special="attachdocument" attrs="{'invisible': [('attachment_number', '&lt;=', 0)]}"/>
```
Comment (ged-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: here, I think that the best way to proceed is to use a <widget> and not a special attribute (a form widget is kind of like a field widget, but not linked to a field)

File: addons/web/static/src/js/views/form/form_renderer.js [hunk 4]
```diff
@@ -2,6 +2,7 @@ odoo.define('web.FormRenderer', function (require) {
 "use strict";
 
 var BasicRenderer = require('web.BasicRenderer');
+var AttachDocument = require('web.Attachment');
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_renderer.js]: by using a <widget>, you should not need to change the form renderer

File: addons/web/static/src/js/views/form/form_controller.js [hunk 5]
```diff
@@ -523,6 +524,22 @@ var FormController = BasicController.extend({
         var state = this.model.get(this.handle);
         this.renderer.confirmChange(state, state.id, [field]);
     },
+    /**
+     * @private
+     * @param {OdooEvent} ev
+     */
+    _onAttachClicked: function (ev) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_controller.js]: this is kind of weird. I think that this should be done in the Attachment widget, if possible

File: addons/web/static/src/js/widgets/attachment.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+     * @param {Widget} parent
+     * @param {Object} params
+     */
+    init: function (parent, params) {
+        this.res_id = params.state.res_id;
+        this.res_model = params.state.model;
+        this.node = params.node;
+        this.state = params.state;
+        this.fileuploadID = _.uniqueId('o_fileupload');
+        return this._super.apply(this, arguments);
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/attachment.js]: no need for return

File: addons/web/static/src/js/widgets/attachment.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        var self = this;
+        $(window).on(self.fileuploadID, self._onFileLoaded.bind(self));
+        return this._super.apply(this, arguments);
+    },
+    /**
+     * @override
+     */
+    destroy: function () {
+        $(window).off(this.fileupload_id);
+        return this._super.apply(this, arguments);
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/attachment.js]: no need for return

File: addons/web/static/src/js/widgets/attachment.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
+odoo.define('web.Attachment', function (require) {
+"use static";
+
+var Widget = require('web.Widget');
+
+var AttachDocument = Widget.extend({
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/attachment.js]: this widget should be added to the widget registry somewhere...

File: addons/hr_expense/views/hr_expense_views.xml [hunk 9]
```diff
@@ -97,7 +97,10 @@
             <field name="arch" type="xml">
                 <xpath expr="/form/*" position="before">
                     <header>
-                        <button name="action_submit_expenses" states="draft" string="Submit to Manager" type="object" class="oe_highlight"/>
+                        <widget name="attach_document" string="Attach Document" action="send_attach_document" attrs="{'invisible': [('attachment_number', '&lt;=', 0)]}"/>
+                        <widget name="attach_document" string="Attach Document" action="send_attach_document" class="oe_highlight" attrs="{'invisible': [('attachment_number', '&gt;=', 1)]}"/>
```
Comment (tivisse) [inline addons/hr_expense/views/hr_expense_views.xml]: I would use the YAGNI principle and set ``action="message_post"`` instead. 

If one day we want to make something more specific, we could use a method instead ;)

**Pull request merged**